.github/workflows/scripts/cherry-pick.sh

```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env bash
+#
+# Tries to cherry-pick merge commits from the main branch to the release branch.
+# Adds comments to the original PRs of the problematic commits.
+#
+set -euo pipefail
+
+MILESTONE="$1"
+BRANCH="$2"
+RELEASE_PATCH="$3"
+
+check_not_empty \
+    MILESTONE \
+    BRANCH \
+    RELEASE_PATCH \
+    \
+    GITHUB_REPOSITORY \
+    DRY_RUN \
+    main_branch
+
+SLACK_MESSAGE_FILE=$(mktemp)
+
+# Increments PICKED variable if managed to pick a cherry.
+cherry_pick() {
+    IFS=$'\t' read -r PR URL COMMIT AUTHOR TITLE <<<"$1"
+
+    # Skip commits merged before branching.
+    if git merge-base --is-ancestor "$COMMIT" HEAD; then
+        gh_log debug "$COMMIT is already on the release branch"
+        return
+    fi
+
+    # Find commits with the specific commit message after the fork point
+    # to not cherry-pick twice the same commit.
+    ALREADY_PICKED_CHERRIES=$(git log \
+        --grep "^(cherry picked from commit $COMMIT)\$" --format='%H' \
+        HEAD..."$FORK_POINT")
+
+    if [ -n "$ALREADY_PICKED_CHERRIES" ]; then
+        gh_summary "Already picked cherries for commit $COMMIT:\n\`\`\`\n$ALREADY_PICKED_CHERRIES\n\`\`\`"
+        return
+    fi
+
+    if git cherry-pick -x "$COMMIT"; then
+        gh_summary "Cherry-picked $COMMIT from PR $PR"
+
+        [ "$DRY_RUN" != "true" ] &&
+            gh pr comment "$PR" --body "Merge commit has been cherry-picked to branch \`$BRANCH\`."
+
+        PICKED=$((PICKED+1))
+    else
+        git cherry-pick --abort
+
+        [ "$DRY_RUN" != "true" ] &&
+            gh pr comment "$PR" --body "Please merge the changes to branch \`$BRANCH\`."
+
+        gh_summary "* [PR $PR]($URL) by **${AUTHOR}** (_${TITLE}_) could not be cherry-picked. Merge commit: \`$COMMIT\`."
+
+        echo "- <$URL|PR $PR> by *$AUTHOR* — $TITLE" \
+            >>"$SLACK_MESSAGE_FILE"
+    fi
+}
+
+find_fork_point() {
+    set +o pipefail
+    diff -u \
+        <(git rev-list --first-parent "$1") \
+        <(git rev-list --first-parent "$2") |
+        sed -ne 's/^ //p' |
+        head -1
+    set -o pipefail
```

Comment:
Would something like this also work here? (I am trying to get rid of the `set +o pipefail`)

```suggestion
    local parent_branch 
    parent_branch="$(git rev-list --first-parent "$1")" || true
    local main_branch
    main_branch="$(git rev-list --first-parent "$2")" || true
    diff -u "$parent_branch" "$main_branch" | 
        sed -ne 's/^ //p' |
        head -1
```

The main question here is actually which pipe do we allow to fail and which not.

Reply:
Understanding question: The `sed -ne 's/^ //p'` does: filter lines starting with a space and print them without that space. Is this correct?

Reply:
Actually, I was thinking entire time that such function should already exist. How about using `git merge-base branch2 branch3`?

Reply:
Your suggested code still needs `set +o pipefail` as it is required to ignore the exit code of `diff`. I guess it could also be done via `if diff ... ; then; fi`.
`git merge-base` didn't always work correctly in my experiments. Teh internets suggested the solution with diff.